Suggestion for call on immutable binding of mutable type

estebank opened the pull request:
When calling a method requiring a mutable self borrow on an inmutable
to a mutable borrow of the type, suggest making the binding mutable.

Fix #83241.

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

estebank commented:
r? @oli-obk

oli-obk commented:
All of the changed tests were a case where the previous suggestion was already applicable, and now still is. While the new version fixes the situation, it always does so by introducing another level of indirection that will immediately get autodereffed.

I think this new suggestion should not apply if the expression is a borrow expression. Also we should probably have the test from the issue in the test suite, as it does cover a case that is not covered by any of the changed tests

Dylan-DPC commented:
@estebank judging by oli's previous comment i'm assuming there are changes expected here

JohnCSimon commented:
Ping from triage:
@estebank build is broken and there are merge conflicts.

estebank commented:
Addressed comments, fixed rebase and fixed structured_suggestion (we currently incorrectly suggest removal of code, which is why this PR originally used `help` instead of `suggestion`).

oli-obk commented:
@bors r+ rollup

camelid commented:
@oli-obk It seems that bors didn't pick up on your approval. Can you try again?

camelid commented:
@bors are you still there?

bors commented:
:cake: Activated

![](https://cloud.githubusercontent.com/assets/1617736/22222924/c07b2a1c-e16d-11e6-91b3-ac659550585c.png)

estebank commented:
@bors r=oli-obk rollup

bors commented:
:pushpin: Commit b252b38a9f071c5ddaf10fb6f0abe7ccf0c3a7d5 has been approved by `oli-obk`

<!-- @bors r=oli-obk b252b38a9f071c5ddaf10fb6f0abe7ccf0c3a7d5 -->
<!-- homu: {"type":"Approved","sha":"b252b38a9f071c5ddaf10fb6f0abe7ccf0c3a7d5","approver":"oli-obk"} -->

estebank commented:
@bors r-

This has old suggestion format, it'll fail to merge.

estebank commented:
@bors r=oli-obk rollup

bors commented:
:pushpin: Commit 5b6f4b94c5679c9cce70d42cb4cbdbd55da2e0f7 has been approved by `oli-obk`

<!-- @bors r=oli-obk 5b6f4b94c5679c9cce70d42cb4cbdbd55da2e0f7 -->
<!-- homu: {"type":"Approved","sha":"5b6f4b94c5679c9cce70d42cb4cbdbd55da2e0f7","approver":"oli-obk"} -->
